sched/semaphore: rework semaphore holder check for priority inheritance

pkarashchenko opened the pull request:
## Summary
Optimize semaphore holder check for priority inheritance

## Impact
Should be none. Code clean-up

## Testing
Passed `ostest` with `spresense:smp` with priority inheritance enabled

pkarashchenko commented:
Just drafted changes, but really didn't have time to run ostest. I would appreciate someone helping with that.

pkarashchenko commented:
Just drafted changes, but really didn't have time to run ostest. I would appreciate someone helping with that.

pkarashchenko commented:
I noticed one strange thing. I tried to run `spresense:smp` with priority inheritance enabled and `CONFIG_SEM_PREALLOCHOLDERS == 0`, but the system crash on start-up. If I enable debug assertions the crash disappear and I can pass `ostest`. @masayuki2009 could you please take a look?

pkarashchenko commented:
It is strange. I just rebuilt the code and can't reproduce the crash. I will do more experiments.

pkarashchenko commented:
Just reproduced the crash
```
[CPU0] _assert: Current Version: NuttX  3.6.1 7b6a424c6d Apr  4 2023 10:56:00 arm
[CPU0] _assert: Assertion failed panic: at file: irq/irq_unexpectedisr.c:54 task(CPU0): CPU0 IDLE 0xd003339
[CPU0] up_dump_register: R0: 2d02bdc4 R1: 2d02bdb8 R2: 00154100  R3: 001540e0
[CPU0] up_dump_register: R4: 2d02bdb8 R5: 0d004861 R6: 00000000  FP: 00000001
[CPU0] up_dump_register: R8: 2d0298dc SB: 2d029a1c SL: 2d029a6c R11: 0d003339
[CPU0] up_dump_register: IP: 00000001 SP: 2d02bd40 LR: 0d00486f  PC: 0d0047b0
[CPU0] up_dump_register: xPSR: a1000000 BASEPRI: 00000080 CONTROL: 00000000
[CPU0] up_dump_register: EXC_RETURN: ffffffe9
[CPU0] dump_stack: IRQ Stack:
[CPU0] dump_stack: sp:     0x2d028d68
[CPU0] dump_stack:   base: 0x2d028650
[CPU0] dump_stack:   size: 00002048
[CPU0] stack_dump: 0x2d028d60: 0d003339 00000000 00000000 7474754e 00000058 00000000 00000000 00000000
[CPU0] stack_dump: 0x2d028d80: 00000000 00000000 00000000 00000000 00000000 00000000 00000000 00000000
[CPU0] stack_dump: 0x2d028da0: 362e3300 0000312e 00000000 00000000 00000000 62370000 32346136 64366334
[CPU0] stack_dump: 0x2d028dc0: 72704120 20342020 33323032 3a303120 303a3635 00000030 00000000 00000000
[CPU0] stack_dump: 0x2d028de0: 00000000 00000000 6d726100 00000000 00000000 00000000 00000000 00000000
[CPU0] stack_dump: 0x2d028e00: 00000000 00000000 2d0297b8 00000003 00000000 2d0298dc 2d029a1c 0d007e17
[CPU0] stack_dump: 0x2d028e20: 00000080 0d00367f 0d003669 0d00363f 00000000 0d001797 0d01cdbc 2d02bdb8
[CPU0] dump_stack: User Stack:
[CPU0] dump_stack: sp:     0x2d02bd40
[CPU0] dump_stack:   base: 0
[CPU0] dump_stack:   size: 00000000
[CPU0] dump_stack: ERROR: User Stack pointer is not within the stack
[CPU0] show_tasks:    PID GROUP   CPU PRI POLICY   TYPE    NPX STATE   EVENT      SIGMASK  STACKBASE  STACKSIZE   COMMAND
[CPU0] show_tasks:   ----   ---  ---- --- -------- ------- --- ------- ---------- -------- 0x2d028650      2048   irq
```

pkarashchenko commented:
And seems like I have the steps to reproduce:
1. `tools/configure.sh spresense:smp`
2. `make menuconfig` -- and enable priority inheritance
3. `make -j8` -- all is working now
4. `make menuconfig` -- and set `CONFIG_SEM_PREALLOCHOLDERS` to `0`
5. `make -j8` -- now I'm getting the crash
6. `make clean`
7. `make -j8` -- now all is working

So it seems like some dependencies are not refreshed with `make menuconfig` and build system is using object files from the previous configuration. I will try to compare binaries and map files after steps 5 and 7.

masayuki2009 commented:
> Just reproduced the crash

@pkarashchenko 
What version of the arm gcc are you using?

pkarashchenko commented:
> > Just reproduced the crash
> 
> @pkarashchenko What version of the arm gcc are you using?

`arm-none-eabi-gcc (Arm GNU Toolchain 12.2.Rel1 (Build arm-12.24)) 12.2.1 20221205`

masayuki2009 commented:
>It is strange. I just rebuilt the code and can't reproduce the crash. I will do more experiments.

@pkarashchenko 
I tried this PR with 

```
--- a/boards/arm/cxd56xx/spresense/configs/smp/defconfig
+++ b/boards/arm/cxd56xx/spresense/configs/smp/defconfig
@@ -40,12 +40,14 @@ CONFIG_NSH_ARCHINIT=y
 CONFIG_NSH_BUILTIN_APPS=y
 CONFIG_NSH_READLINE=y
 CONFIG_PREALLOC_TIMERS=4
+CONFIG_PRIORITY_INHERITANCE=y
 CONFIG_RAM_SIZE=1572864
 CONFIG_RAM_START=0x0d000000
 CONFIG_READLINE_CMD_HISTORY=y
 CONFIG_RR_INTERVAL=200
 CONFIG_RTC=y
 CONFIG_RTC_DRIVER=y
+CONFIG_SEM_PREALLOCHOLDERS=0
 CONFIG_SMP=y
 CONFIG_SMP_NCPUS=2
 CONFIG_SPI=y

$ arm-none-eabi-gcc --version
arm-none-eabi-gcc (Arm GNU Toolchain 12.2.MPACBTI-Rel1 (Build arm-12-mpacbti.34)) 12.2.1 20230214
Copyright (C) 2022 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
```

And the ostest passed without any errors.

```
NuttShell (NSH) NuttX-3.6.1
nsh> uname -a
NuttX 3.6.1 7b6a424c6d-dirty Apr  4 2023 17:35:14 arm spresense
nsh> ps
  PID GROUP CPU PRI POLICY   TYPE    NPX STATE    EVENT     SIGMASK           STACK COMMAND
    0     0   0   0 FIFO     Kthread N-- Assigned           0000000000000000 001000 CPU0 IDLE
    1     1   1   0 FIFO     Kthread N-- Running            0000000000000000 001000 CPU1 IDLE
    3     3 --- 200 RR       Task    --- Waiting  MQ empty  0000000000000000 000976 cxd56_pm_task
    4     4   0 100 RR       Task    --- Running            0000000000000000 002000 spresense_main
nsh> free
                   total       used       free    largest  nused  nfree
        Umem:    1392840      15744    1377096    1374864     77      3
nsh> ostest
stdio_test: write fd=1
stdio_test: Standard I/O Check: printf
stdio_test: write fd=2
...
user_main: priority inheritance test
priority_inheritance: Started
priority_inheritance: Starting lowpri_thread-1 (of 1) at 1
priority_inheritance: Set lowpri_thread-1 priority to 1
priority_inheritance: Waiting...
lowpri_thread-1: Started
lowpri_thread-1: initial priority: 1
lowpri_thread-1: Waiting for the midle pri task to run
    g_middlestate:  0
    g_highstate[0]: 0
    I still have a count on the semaphore
lowpri_thread-1: Waiting for the midle pri task to run
    g_middlestate:  0
    g_highstate[0]: 0
    I still have a count on the semaphore
priority_inheritance: Starting medpri_thread at 99
priority_inheritance: Set medpri_thread priority to 99
priority_inheritance: Waiting...
medpri_thread: Started ... I won't let go of the CPU!
lowpri_thread-1: Sem count: 0, No. highpri thread: 0
lowpri_thread-1: SUCCESS priority before sem_post: 1
lowpri_thread-1: SUCCESS final priority: 1
lowpri_thread-1: Okay... I'm done!
priority_inheritance: Starting highpri_thread-1 (of 1) at 255
priority_inheritance: Set highpri_thread-1 priority to 255
priority_inheritance: Waiting for highpri_thread-1 to complete
highpri_thread-1: Started
highpri_thread-1: Calling sem_wait()
highpri_thread-1: SUCCESS midpri_thread is still running!
highpri_thread-1: Okay... I'm done!
priority_inheritance: Waiting for medpri_thread to complete
medpri_thread: Okay... I'm done!
priority_inheritance: Waiting for lowpri_thread-1 to complete
priority_inheritance: Restoration Test:
priority_inheritance: Task0 initial priority is:200
priority_inheritance: Task1 initial priority is:210
priority_inheritance: Waiting for Task-0 to complete
priority_inheritance: Task2 initial priority is:220
priority_inheritance: Task0 Started, waiting 0 uS to take count
priority_inheritance: Task1 Started, waiting 10000 uS to take count
priority_inheritance: Task2 Started, waiting 20000 uS to take count
priority_inheritance: Task0 Posted
priority_inheritance: Task0 priority was:200 is:200
priority_inheritance: Waiting for Task-1 to complete
priority_inheritance: Task2 Posted
priority_inheritance: Task2 priority was:220 is:220
priority_inheritance: Task1 Posted
priority_inheritance: Task1 priority was:210 is:210
priority_inheritance: Waiting for Task-2 to complete
priority_inheritance: PASSED Priority were correctly restored.
priority_inheritance: Finished

End of test memory usage:
VARIABLE  BEFORE   AFTER
======== ======== ========
arena      1540c8   1540c8
ordblks        11       12
mxordblk   143ed8   143ed8
uordblks     8520     a828
fordblks   14bba8   1498a0

user_main: vfork() test
vfork_test: Child 94 ran successfully

Final memory usage:
VARIABLE  BEFORE   AFTER
======== ======== ========
arena      1540c8   1540c8
ordblks         4       12
mxordblk   14b720   143ed8
uordblks     74e8     c028
fordblks   14cbe0   1480a0
user_main: Exiting
ostest_main: Exiting with status 0
```

pkarashchenko commented:
@masayuki2009 seems like we have some issue with the build system. The map file generated after `make menuconfig` and after `make clean` with the sam config are different (and binaries as well)

pkarashchenko commented:
@xiaoxiang781216 do you have any idea how we can debug this? I mean is that an issue of a build system or some interference with compiler version.

pkarashchenko commented:
I will fill a ticket related to build problem. Anyway it is not directly related to this PR

davids5 commented on sched/semaphore/sem_holder.c:
```diff
@@ -103,20 +103,18 @@ nxsem_allocholder(FAR sem_t *sem, FAR struct tcb_s *htcb)
-      pholder->sem    = sem;
-      pholder->htcb   = htcb;
-      pholder->counts = 0;
+  DEBUGASSERT(pholder != NULL);
```
If we can run out of holder 2 ways here. Why leave this debug detection only? The following code will corrupt memory. Not having enough holders, if it were survivable will still cause PI issues,

pkarashchenko replied:
I just didn't want to change existing behaviour that use `DEBUGPANIC();` and not `PANIC();`. But I agree that `ASSERT(pholder != NULL);` might be better here.

pkarashchenko replied:
Done

davids5 reviewed (commented):
I wonder if we want to allow a misconfigured system to run at all. If we run out of holders. The PI is going to silently happen. Should it just PANIC?

davids5 reviewed (approved):
Looks good. We should light breaking change to give a heads up that misconfigured system will Panic.

xiaoxiang781216 commented:
> @xiaoxiang781216 do you have any idea how we can debug this? I mean is that an issue of a build system or some interference with compiler version.

I never expect that the incremental build will work after you change config. My suggestion is always to clean and rebuild the image in this case.

xiaoxiang781216 commented:
Please revert https://github.com/apache/nuttx/pull/8938 and apply your change after it.

pkarashchenko commented on sched/semaphore/sem_holder.c:
```diff
@@ -102,21 +102,17 @@ nxsem_allocholder(FAR sem_t *sem, FAR struct tcb_s *htcb)
-      pholder->sem    = sem;
-      pholder->htcb   = htcb;
-      pholder->counts = 0;
+  pholder->sem    = sem;
```
Done

pkarashchenko commented:
> Please revert #8938 and apply your change after it.

Done

pkarashchenko commented:
> > @xiaoxiang781216 do you have any idea how we can debug this? I mean is that an issue of a build system or some interference with compiler version.
> 
> I never expect that the incremental build will work after you change config. My suggestion is always to clean and rebuild the image in this case.

Ok. Makes sense. I'll close the issue. Most of the code is rebuilt because of `#include <nuttx/config.h>`, but there are some files that do not include it directly, so stale objects are used.

xiaoxiang781216 commented on sched/semaphore/sem_holder.c:
```diff
@@ -891,6 +789,8 @@ void nxsem_restore_baseprio(FAR struct tcb_s *stcb, FAR sem_t *sem)
               (sem->semcount <= 0 && stcb != NULL));
 #endif
 
+  DEBUGASSERT(!up_interrupt_context());
```
why can't be called from the interrupt context?

pkarashchenko replied:
The same as `nxsem_checkholder` -> `if (up_interrupt_context()) { return false; }`

xiaoxiang781216 commented on sched/semaphore/sem_holder.c:
```diff
@@ -792,61 +697,54 @@ void nxsem_boost_priority(FAR sem_t *sem)
+      FAR struct semholder_s *pholder;
 
-  /* Find the container for this holder */
+      DEBUGASSERT(!up_interrupt_context());
```
ditto

pkarashchenko replied:
The same as `nxsem_checkholder` -> `if (up_interrupt_context()) { return false; }`

pkarashchenko replied:
In general we can't post semaphore with priority inheritance enabled from the interrupt

xiaoxiang781216 replied:
Yes, I think so. After we add more check, the semaphore with priority inheritance is more difficult to use in non-lock case.

pkarashchenko replied:
The question is only where we should check it. Semaphore holders APIs are internal APIs and interrupt context check may be done once in nxsem_post instead of two times like in this PR. On the other hand the current approach will catch possible misuse of semaphore holders APIs.

xiaoxiang781216 commented on sched/semaphore/sem_holder.c:
```diff
@@ -792,61 +697,54 @@ void nxsem_boost_priority(FAR sem_t *sem)
+   * inheritance is effectively disabled.
+   */
+
+  if (!is_idle_task(rtcb) && prioinherit == SEM_PRIO_INHERIT)
```
why skip in the idle task?

pkarashchenko commented:
@xiaoxiang781216 please take a look if your question are answered. Thanks!

pkarashchenko commented:
@patacongo may I get your approval on this change?

acassis commented:
> Current Version: NuttX  3.6.1

Any idea why it is happening? Seems there is some issue in our current script and sometimes when we aren't in the upstream branch it get confused.

pkarashchenko commented:
> > Current Version: NuttX  3.6.1
> 
> Any idea why it is happening? Seems there is some issue in our current script and sometimes when we aren't in the upstream branch it get confused.

I have no idea, but this is what I get from building the latest master

davids5 commented:
@acassis Fetch the tags.

davids5 commented:
@pkarashchenko We ran in to an issue with an older version of the PI code and I am wondering if it exits in the latest code?

sched_lock() is not respected on a call to sem_post that reprioritizes 

Causing code like following to not work
```
          sched_lock();
          nxsem_post(&dev->d_bfsem);   ------------>Swapped out if there is a higher priority waiter. 
          ret = nxsem_wait(&dev->d_wrsem);
          sched_unlock();
```
Has this been fixed? If so, can you point me to the mitigation?

pkarashchenko commented:
@davids5 I expect this to be fixed by https://github.com/apache/nuttx/pull/7464 and later optimised by https://github.com/apache/nuttx/pull/9039
But above example will be working only in non-SMP case since `sched_lock()/sched_unlock()` will not work if threads are running on the different cores. Or you need to add `pthread_attr_setaffinity_np()` for two pthreads and bind them to the same CPU.
I will prioritise addressing all comments from https://github.com/apache/nuttx-apps/pull/1711 this week as all prerequisites seems to be met.
